tagmanager/ctags/powershell.c

```diff
@@ -0,0 +1,658 @@
+/*
+*   Copyright (c) 2015, Enrico Tröger <enrico.troeger@uvena.de>
+*
+*   Loosely based on the PHP tags parser since the syntax is somewhat similar
+*   regarding variable and function definitions.
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License.
+*
+*   This module contains code for generating tags for Windows PowerShell scripts.
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"  /* must always come first */
+#include "main.h"
+#include "parse.h"
+#include "read.h"
+#include "vstring.h"
+#include "keyword.h"
+#include "entry.h"
+#include <string.h>
+
+#define SCOPE_SEPARATOR ":"
+
+
+#define ACCESS_UNDEFINED "undefined"
+static const char *const accessTypes[] = {
+	ACCESS_UNDEFINED,
+	"global",
+	"local",
+	"script",
+	"private",
+	NULL
+};
+
+typedef enum {
+	K_FUNCTION,
+	K_VARIABLE,
+	COUNT_KIND
+} powerShellKind;
+
+static kindOption PowerShellKinds[COUNT_KIND] = {
+	{ TRUE, 'f', "function",	"functions" },
+	{ TRUE, 'v', "variable",	"variables" }
+};
+
+
+typedef enum eTokenType {
+	TOKEN_UNDEFINED,
+	TOKEN_EOF,
+	TOKEN_CHARACTER,
+	TOKEN_CLOSE_PAREN,
+	TOKEN_SEMICOLON,
+	TOKEN_COLON,
+	TOKEN_COMMA,
+	TOKEN_KEYWORD,
+	TOKEN_OPEN_PAREN,
+	TOKEN_OPERATOR,
+	TOKEN_IDENTIFIER,
+	TOKEN_STRING,
+	TOKEN_PERIOD,
+	TOKEN_OPEN_CURLY,
+	TOKEN_CLOSE_CURLY,
+	TOKEN_EQUAL_SIGN,
+	TOKEN_OPEN_SQUARE,
+	TOKEN_CLOSE_SQUARE,
+	TOKEN_VARIABLE
+} tokenType;
+
+typedef struct {
+	tokenType		type;
+	vString *		string;
+	vString *		scope;
+	unsigned long 	lineNumber;
+	MIOPos			filePosition;
+	int 			parentKind; /* -1 if none */
+} tokenInfo;
+
+
+static const char *const findValidAccessType (const char *const access)
+{
+	int i = 0;
+	while (accessTypes[i] != NULL)
+	{
+		if (strcasecmp (access, accessTypes[i]) == 0)
+			return accessTypes[i];
+		i++;
+	}
+	return ACCESS_UNDEFINED;
+}
+
+static void initPowerShellEntry (tagEntryInfo *const e, const tokenInfo *const token,
+								 const powerShellKind kind, const char *const access)
+{
+	static vString *fullScope = NULL;
```

Comment:
Good catch. I copied this from the PHP parser :).
Having a look at the fishman's repo code, they made it global in the PHP parser and free it properly, done so for PowerShell as well in 5057206